### DISCUSSION TITLE
Error on initial page load

### DISCUSSION THREAD

**Discussion by evilkermit** [Kitware/trame#624] [Q&A] [answered]
Hi there!

We have a few Trame applications that are deployed through https://galaxyproject.org/, and we are seeing the following error pretty consistently when each Trame application is first launched. If it's helpful, adding a fairly long delay before attempting to open the application does avoid this error.

![connection_error](https://github.com/user-attachments/assets/34f7c3d4-404e-473c-b626-7f9757090fd5)

It seems like the loading screen for Trame becomes available to load before a websocket connection is possible. After the error happens, the page gets stuck on this loading screen until the user refreshes.

Would it be possible for the Trame client to either 1. refresh the page automatically if the websocket connection fails, or 2. retry the websocket connection periodically from the loading screen if the websocket connection fails? If not, do you have any recommendations on how we should handle this (maybe there's a clean way for us to check if the websocket is ready before we tell the user the application is available)?

**Comment by jourdain** [comment]
What you are suggesting seems to be a workaround more than a solution. As I don't know how the process start for your trame application, it would be hard for me to properly advise. But in general, you should not connect the client until the server is ready unless the launcher is used.

One possible option but I'm not sure it will work for your usecase is to add `?reconnect=auto` in the URL.

**Reply by evilkermit** [reply]
Thank you for the quick response! The short version is that we build Docker images for each Trame app, and Galaxy runs the images and manages a proxy for navigating a user to their launched instance of the Trame app. I can try to put together a complete reproduction if needed, but that's a bit difficult due to Galaxy.

Is there a general mechanism for determining if the Trame server is completely ready from external software (for example, is there some way I could determine if the server is ready via curl)? If so, I think we could adapt a solution from that. Also, setting reconnect=auto doesn't seem to work. I think the error in the screenshot interrupts execution of the main JS script, so the page is just dead at that point.

**Reply by jourdain** [reply]
If the docker is started on demand, then you do not need the launcher setup from the docker bundle. And you can either monitor the status of the job by looking at its output or if the process respond to http request (if you skip the launcher part).

I guess it might be better to chat about it.

**Reply by evilkermit** [reply, marked as answer]
Sorry for the delay, I was working on getting a reproduction to share and in the process learned enough about it to realize what the underlying issue was. Not very surprising, but it doesn't have anything to do with Trame itself. I'll detail below what I learned in case it's useful to someone else!

When we were launching our Docker images we were starting nginx to serve a static bundle of the Trame front-end and then after that were starting the Trame server. Because nginx was already available, our system was reporting the application as available before the Trame server had completed startup. Deferring the nginx startup until after we get a successful response from the Trame server seems to be sufficient to resolve the issue.
